In [442]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import StandardScaler
from sklearn import metrics as met
from statistics import mean
from scipy.stats import linregress

### House Training processing

In [465]:
house_training = pd.read_csv('DS_4400_Dataset_Spring_2022/train.csv')
house_training = house_training.drop('Unnamed: 0', axis=1)
house_training

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0    221900.0         3       1.00         1180      5650     1.0           0   
1    538000.0         3       2.25         2570      7242     2.0           0   
2    180000.0         2       1.00          770     10000     1.0           0   
3    604000.0         4       3.00         1960      5000     1.0           0   
4    510000.0         3       2.00         1680      8080     1.0           0   
..        ...       ...        ...          ...       ...     ...         ...   
995  291000.0         4       2.50         1860      6325     2.0           0   
996  199950.0         2       2.75         1590     20917     1.5           0   
997  553500.0         2       1.00          850      2340     1.0           0   
998  189950.0         2       1.00         1030      4188     1.0           0   
999  289000.0         3       2.00         1850      9550     1.0           0   

     view  condition  grade  sqft_above  sqft_basement  yr_built  \
0       0          3      7        1180              0      1955   
1       0          3      7        2170            400      1951   
2       0          3      6         770              0      1933   
3       0          5      7        1050            910      1965   
4       0          3      8        1680              0      1987   
..    ...        ...    ...         ...            ...       ...   
995     0          4      7        1860              0      1991   
996     0          3      5        1590              0      1920   
997     0          3      7         850              0      1922   
998     0          3      8        1030              0      1981   
999     0          3      8        1850              0      1988   

     yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
0               0    98178  47.5112 -122.257           1340        5650  
1            1991    98125  47.7210 -122.319           1690        7639  
2               0    98028  47.7379 -122.233           2720        8062  
3               0    98136  47.5208 -122.393           1360        5000  
4               0    98074  47.6168 -122.045           1800        7503  
..            ...      ...      ...      ...            ...         ...  
995             0    98038  47.3492 -122.030           1860        6449  
996             0    98001  47.2786 -122.250           1310        6000  
997             0    98105  47.6707 -122.328           1300        3000  
998             0    98038  47.3738 -122.057           1450        3376  
999             0    98001  47.3225 -122.273           2250        9550  

[1000 rows x 19 columns]

In [466]:
house_training

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0    221900.0         3       1.00         1180      5650     1.0           0   
1    538000.0         3       2.25         2570      7242     2.0           0   
2    180000.0         2       1.00          770     10000     1.0           0   
3    604000.0         4       3.00         1960      5000     1.0           0   
4    510000.0         3       2.00         1680      8080     1.0           0   
..        ...       ...        ...          ...       ...     ...         ...   
995  291000.0         4       2.50         1860      6325     2.0           0   
996  199950.0         2       2.75         1590     20917     1.5           0   
997  553500.0         2       1.00          850      2340     1.0           0   
998  189950.0         2       1.00         1030      4188     1.0           0   
999  289000.0         3       2.00         1850      9550     1.0           0   

     view  condition  grade  sqft_above  sqft_basement  yr_built  \
0       0          3      7        1180              0      1955   
1       0          3      7        2170            400      1951   
2       0          3      6         770              0      1933   
3       0          5      7        1050            910      1965   
4       0          3      8        1680              0      1987   
..    ...        ...    ...         ...            ...       ...   
995     0          4      7        1860              0      1991   
996     0          3      5        1590              0      1920   
997     0          3      7         850              0      1922   
998     0          3      8        1030              0      1981   
999     0          3      8        1850              0      1988   

     yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
0               0    98178  47.5112 -122.257           1340        5650  
1            1991    98125  47.7210 -122.319           1690        7639  
2               0    98028  47.7379 -122.233           2720        8062  
3               0    98136  47.5208 -122.393           1360        5000  
4               0    98074  47.6168 -122.045           1800        7503  
..            ...      ...      ...      ...            ...         ...  
995             0    98038  47.3492 -122.030           1860        6449  
996             0    98001  47.2786 -122.250           1310        6000  
997             0    98105  47.6707 -122.328           1300        3000  
998             0    98038  47.3738 -122.057           1450        3376  
999             0    98001  47.3225 -122.273           2250        9550  

[1000 rows x 19 columns]

In [467]:
train_price = house_training['price'] / 1000
#house_training1['price'] = price_adj
#house_training1
train_price = train_price.tolist()

In [468]:
#house_training = house_training.drop(['zipcode', 'price'], axis=1)
house_training

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0    221900.0         3       1.00         1180      5650     1.0           0   
1    538000.0         3       2.25         2570      7242     2.0           0   
2    180000.0         2       1.00          770     10000     1.0           0   
3    604000.0         4       3.00         1960      5000     1.0           0   
4    510000.0         3       2.00         1680      8080     1.0           0   
..        ...       ...        ...          ...       ...     ...         ...   
995  291000.0         4       2.50         1860      6325     2.0           0   
996  199950.0         2       2.75         1590     20917     1.5           0   
997  553500.0         2       1.00          850      2340     1.0           0   
998  189950.0         2       1.00         1030      4188     1.0           0   
999  289000.0         3       2.00         1850      9550     1.0           0   

     view  condition  grade  sqft_above  sqft_basement  yr_built  \
0       0          3      7        1180              0      1955   
1       0          3      7        2170            400      1951   
2       0          3      6         770              0      1933   
3       0          5      7        1050            910      1965   
4       0          3      8        1680              0      1987   
..    ...        ...    ...         ...            ...       ...   
995     0          4      7        1860              0      1991   
996     0          3      5        1590              0      1920   
997     0          3      7         850              0      1922   
998     0          3      8        1030              0      1981   
999     0          3      8        1850              0      1988   

     yr_renovated  zipcode      lat     long  sqft_living15  sqft_lot15  
0               0    98178  47.5112 -122.257           1340        5650  
1            1991    98125  47.7210 -122.319           1690        7639  
2               0    98028  47.7379 -122.233           2720        8062  
3               0    98136  47.5208 -122.393           1360        5000  
4               0    98074  47.6168 -122.045           1800        7503  
..            ...      ...      ...      ...            ...         ...  
995             0    98038  47.3492 -122.030           1860        6449  
996             0    98001  47.2786 -122.250           1310        6000  
997             0    98105  47.6707 -122.328           1300        3000  
998             0    98038  47.3738 -122.057           1450        3376  
999             0    98001  47.3225 -122.273           2250        9550  

[1000 rows x 19 columns]

In [469]:
norm_col = [1] * len(house_training)

In [470]:
house_training.insert(loc=0, column='norm', value=norm_col)

In [471]:
house_training = house_training.drop(['zipcode', 'price'], axis=1)

In [472]:
house_training

norm  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0       1         3       1.00         1180      5650     1.0           0   
1       1         3       2.25         2570      7242     2.0           0   
2       1         2       1.00          770     10000     1.0           0   
3       1         4       3.00         1960      5000     1.0           0   
4       1         3       2.00         1680      8080     1.0           0   
..    ...       ...        ...          ...       ...     ...         ...   
995     1         4       2.50         1860      6325     2.0           0   
996     1         2       2.75         1590     20917     1.5           0   
997     1         2       1.00          850      2340     1.0           0   
998     1         2       1.00         1030      4188     1.0           0   
999     1         3       2.00         1850      9550     1.0           0   

     view  condition  grade  sqft_above  sqft_basement  yr_built  \
0       0          3      7        1180              0      1955   
1       0          3      7        2170            400      1951   
2       0          3      6         770              0      1933   
3       0          5      7        1050            910      1965   
4       0          3      8        1680              0      1987   
..    ...        ...    ...         ...            ...       ...   
995     0          4      7        1860              0      1991   
996     0          3      5        1590              0      1920   
997     0          3      7         850              0      1922   
998     0          3      8        1030              0      1981   
999     0          3      8        1850              0      1988   

     yr_renovated      lat     long  sqft_living15  sqft_lot15  
0               0  47.5112 -122.257           1340        5650  
1            1991  47.7210 -122.319           1690        7639  
2               0  47.7379 -122.233           2720        8062  
3               0  47.5208 -122.393           1360        5000  
4               0  47.6168 -122.045           1800        7503  
..            ...      ...      ...            ...         ...  
995             0  47.3492 -122.030           1860        6449  
996             0  47.2786 -122.250           1310        6000  
997             0  47.6707 -122.328           1300        3000  
998             0  47.3738 -122.057           1450        3376  
999             0  47.3225 -122.273           2250        9550  

[1000 rows x 18 columns]

In [473]:
# remove mean and fit to variance
train_scaler = StandardScaler()
# fit training data to scaler
train_scaler.fit(house_training)
# perform standardization by scaling
x_train = train_scaler.transform(house_training)

In [474]:
x_train

array([[ 0.        , -0.40982347, -1.44988843, ..., -0.35519332,
        -0.96563661, -0.3128578 ],
       [ 0.        , -0.40982347,  0.28318404, ..., -0.7998304 ,
        -0.44332966, -0.23355563],
       [ 0.        , -1.58410276, -1.44988843, ..., -0.18307573,
         1.09374507, -0.21669046],
       ...,
       [ 0.        , -1.58410276, -1.44988843, ..., -0.86437449,
        -1.02532883, -0.4185143 ],
       [ 0.        , -1.58410276, -1.44988843, ...,  1.07911986,
        -0.80148299, -0.40352304],
       [ 0.        , -0.40982347, -0.06343045, ..., -0.46993837,
         0.39236146, -0.15736334]])

### House Testing processing

In [475]:
house_testing = pd.read_csv('DS_4400_Dataset_Spring_2022/test.csv')
house_testing = house_testing.drop('Unnamed: 0', axis=1)
house_testing

id             date     price  bedrooms  bathrooms  sqft_living  \
0    3131200640  20150427T000000  700000.0         4       2.00         1830   
1     984000710  20141022T000000  270000.0         3       2.00         1560   
2    4167300350  20140508T000000  258000.0         4       1.75         1730   
3    2826049282  20140614T000000  530000.0         3       2.50         1930   
4    8946750030  20141218T000000  245000.0         3       2.25         1422   
..          ...              ...       ...       ...        ...          ...   
995  1022059082  20140508T000000  307000.0         3       1.75         1890   
996   179000350  20141105T000000  194000.0         3       1.50         1010   
997  3915500045  20141114T000000  180000.0         3       1.00         1010   
998  3578400030  20140718T000000  465000.0         4       2.25         2340   
999  9346700320  20150323T000000  722500.0         4       2.50         2460   

     sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
0        4590     2.0           0     0  ...      8        1830   
1        8853     1.0           0     0  ...      7        1560   
2        8320     1.0           0     0  ...      7        1230   
3        7214     2.0           0     0  ...      8        1930   
4        3677     2.0           0     0  ...      7        1422   
..        ...     ...         ...   ...  ...    ...         ...   
995     13860     1.0           0     0  ...      7        1890   
996      5000     1.0           0     0  ...      6        1010   
997      8581     1.0           0     0  ...      6        1010   
998     13383     1.0           0     0  ...      8        1170   
999      9296     2.0           0     0  ...      9        2460   

     sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0                0      1908             0    98105  47.6593 -122.327   
1                0      1967             0    98058  47.4312 -122.171   
2              500      1977             0    98023  47.3270 -122.361   
3                0      2005             0    98125  47.7191 -122.309   
4                0      2012             0    98092  47.3204 -122.178   
..             ...       ...           ...      ...      ...      ...   
995              0      1966             0    98042  47.4156 -122.149   
996              0      1943             0    98178  47.4925 -122.278   
997              0      1920             0    98002  47.3043 -122.216   
998           1170      1983             0    98074  47.6211 -122.037   
999              0      1978             0    98007  47.6125 -122.152   

     sqft_living15  sqft_lot15  
0             1650        4590  
1             1610        8750  
2             1840        9800  
3             1930        7266  
4             1677        3677  
..             ...         ...  
995           1500       14536  
996            980        5000  
997           1060       10354  
998           1810       12532  
999           2730        9900  

[1000 rows x 21 columns]

In [476]:
house_testing = house_testing.drop(['id', 'date', 'zipcode'], axis=1)

In [477]:
house_testing

price  bedrooms  bathrooms  sqft_living  sqft_lot  floors  waterfront  \
0    700000.0         4       2.00         1830      4590     2.0           0   
1    270000.0         3       2.00         1560      8853     1.0           0   
2    258000.0         4       1.75         1730      8320     1.0           0   
3    530000.0         3       2.50         1930      7214     2.0           0   
4    245000.0         3       2.25         1422      3677     2.0           0   
..        ...       ...        ...          ...       ...     ...         ...   
995  307000.0         3       1.75         1890     13860     1.0           0   
996  194000.0         3       1.50         1010      5000     1.0           0   
997  180000.0         3       1.00         1010      8581     1.0           0   
998  465000.0         4       2.25         2340     13383     1.0           0   
999  722500.0         4       2.50         2460      9296     2.0           0   

     view  condition  grade  sqft_above  sqft_basement  yr_built  \
0       0          3      8        1830              0      1908   
1       0          3      7        1560              0      1967   
2       0          3      7        1230            500      1977   
3       0          3      8        1930              0      2005   
4       0          3      7        1422              0      2012   
..    ...        ...    ...         ...            ...       ...   
995     0          5      7        1890              0      1966   
996     0          3      6        1010              0      1943   
997     0          4      6        1010              0      1920   
998     0          3      8        1170           1170      1983   
999     0          3      9        2460              0      1978   

     yr_renovated      lat     long  sqft_living15  sqft_lot15  
0               0  47.6593 -122.327           1650        4590  
1               0  47.4312 -122.171           1610        8750  
2               0  47.3270 -122.361           1840        9800  
3               0  47.7191 -122.309           1930        7266  
4               0  47.3204 -122.178           1677        3677  
..            ...      ...      ...            ...         ...  
995             0  47.4156 -122.149           1500       14536  
996             0  47.4925 -122.278            980        5000  
997             0  47.3043 -122.216           1060       10354  
998             0  47.6211 -122.037           1810       12532  
999             0  47.6125 -122.152           2730        9900  

[1000 rows x 18 columns]

In [480]:
# remove mean and fit to variance
train_scaler = StandardScaler()
# fit training data to scaler
train_scaler.fit(house_testing)
# perform standardization by scaling
x_test = train_scaler.transform(house_testing)
# price from testing
test_price = house_testing['price'] / 1000

In [481]:
test_price

0      700.0
1      270.0
2      258.0
3      530.0
4      245.0
       ...  
995    307.0
996    194.0
997    180.0
998    465.0
999    722.5
Name: price, Length: 1000, dtype: float64

In [482]:
x_test

array([[ 0.38774154,  0.61971096, -0.12765819, ..., -0.83161535,
        -0.49282611, -0.32497811],
       [-0.66534862, -0.45245333, -0.12765819, ...,  0.32371455,
        -0.55042162, -0.17705032],
       [-0.69473718,  0.61971096, -0.44286359, ..., -1.08341802,
        -0.21924741, -0.13971278],
       ...,
       [-0.88576284, -0.45245333, -1.38847979, ..., -0.00955369,
        -1.34235996, -0.12001278],
       [-0.18778448,  0.61971096,  0.18754721, ...,  1.31611331,
        -0.26244405, -0.04256405],
       [ 0.4428451 ,  0.61971096,  0.50275261, ...,  0.46442781,
         1.06225281, -0.13615682]])

How do we account for them being different sizes?

## Problem 2

In [484]:
# initializes Linear Regression
reg = LinearRegression()

# fits the training and response variable
reg1 = reg.fit(x_train, train_price)

# predict response with training data
pred_train = reg1.predict(x_train)

# print weights of features
print(reg.coef_)

[ 0.00000000e+00 -1.28709298e+01  1.80756277e+01 -7.12958833e+14
  1.08965572e+01  7.92330091e+00  6.36090289e+01  4.83003123e+01
  1.31525487e+01  9.20166150e+01  6.34389208e+14  3.62046708e+14
 -6.75328079e+01  1.72666911e+01  7.85468253e+01 -9.29685175e-01
  4.56244314e+01 -1.29506561e+01]


In [485]:
# initializes Linear Regression
reg = LinearRegression()

# fits the training
reg1 = reg.fit(x_test, test_price)

#for column in house_train_n:
pred_test = reg1.predict(x_test)
#y_pred_list.append(pred_y)
    
#y_pred_list
print(reg.coef_)

[ 4.08322113e+02 -8.78051473e-14 -4.56725437e-14 -1.71352669e-14
  3.31841353e-14  1.14570501e-13  1.57654181e-13 -5.55762281e-14
 -6.33576669e-15  5.54782857e-14  1.89123062e-13  9.96137895e-14
  5.19532821e-14  1.07104955e-13  3.80657690e-14 -8.54423195e-14
  8.05097504e-14  3.82866244e-14]


In [486]:
# compute r^2 score with training 
train_r_sqr_score = met.r2_score(y_true=train_price, y_pred=pred_train)
# compute mean squared error
train_mse = met.mean_squared_error(train_price, pred_train)

# print score and mse
print("r^2 score: " + f'{train_r_sqr_score}')
print("MSE: " + f'{train_mse}')

r^2 score: 0.7265411845309646
MSE: 31485.275156391483


In [487]:
# compute r^2 score with training
test_r_sqr_score = met.r2_score(y_true=test_price, y_pred=pred_test)
# compute mean squared error
test_mse = met.mean_squared_error(test_price, pred_test)

# print score and mse
print("r^2 score: " + f'{test_r_sqr_score}')
print("MSE: " + f'{test_mse}')

r^2 score: 1.0
MSE: 8.691778000985667e-26


Because r^2 score is a measure of the extent that the variance in the dependent response can be explained by the independent training variable, having a higher r^2 score means the model can predict the variations of the test data to a greater degree. A score of ~0.7265 indicates that a fairly large proportion of the variation can be accounted for by the model, meaning this model is a fairly good predictor for new test points. The large MSE value means that the variance and bias in the model is large, because MSE describes the difference between predicted points and the average value mean. For the testing data, the r^2 score is 1.0, indicating that the model fits the data perfectly and has the ability to absolutely predict new test points. The test MSE is very small, indicating that this predictor has very little bias and the error is small.

## Problem 3

In [488]:
def closed_form_sol(df_, y_):
    '''
    Produces the theta weight of trained data using closed form solution
    
    df_ (DataFrame): dataframe of training data
    y_ (list): list of points of the feature to be predicted
    x_samp (list): list of values for each feature
    
    Return:
    (np.array): array of trained weights (theta) for each feature in dataframe
    
    '''
    
    # turns given df to numpy array
    matrix = df_
    # turns given y_pred list to numpy array
    y_mat = y_
    
    # transposes the df_matrix
    matrix_t = np.transpose(df_)
    
    # performs close form operations to produce list of thetas
    cfm1 = np.linalg.pinv(np.matmul(df_, matrix_t))
    cfm2 = np.matmul(matrix_t, cfm1)
    cfm3 = np.matmul(cfm2, y_)
    
    return cfm3

In [489]:
# test closed_form_sol on training set and training price
train_cfm = closed_form_sol(x_train, train_price)
train_cfm

array([  0.        , -12.52196187,  18.52763251,  56.7488368 ,
        10.88186845,   8.04372084,  63.74289956,  48.20010852,
        12.96426936,  92.23147482,  48.29008886,  27.13703247,
       -67.64311741,  17.27137953,  78.37573693,  -1.03520308,
        45.57765781, -12.93009098])

In [490]:
# test closed_form_sol on training set and training price
test_cfm = closed_form_sol(x_test, test_price)
test_cfm

array([ 4.08322113e+02, -9.30810984e-13,  1.10844667e-12,  4.11048973e-12,
        1.34292577e-12, -3.27204930e-12,  8.70414851e-13, -6.83897383e-14,
        2.72504241e-12, -4.77484718e-12,  6.41620090e-12, -3.60245167e-12,
        1.03952402e-11,  1.61648472e-13,  1.59849911e-11, -2.19948504e-11,
        5.09459142e-12,  4.51549909e-12])

In [491]:
train_cfm_r = np.matmul(x_train, closed_form_sol(x_train, train_price))
train_cfm_r

array([-2.49008272e+02,  2.08703650e+02, -9.27576491e+01, -9.42151019e+01,
       -6.99382127e+01,  8.60376147e+02, -2.80540658e+02, -3.05723979e+02,
       -1.56094471e+02, -2.31285386e+02,  2.48935386e+02, -8.85642924e+01,
        4.29091975e+01, -1.84160200e+02,  5.62809036e+01,  4.03480322e+02,
       -5.86812172e+01,  1.32570807e+01, -2.60865334e+02, -3.56776175e+02,
       -4.14116309e+01,  6.15858933e+02, -1.46538834e+02, -4.03385919e+02,
       -3.45822876e+01, -2.93023889e+02,  2.19395276e+02,  1.10456261e+02,
       -5.67905568e+01,  1.19327560e+02,  1.06166317e+01, -1.84961465e+02,
        9.14370442e+01,  2.24545512e+01, -1.73791722e+02,  2.76251719e+02,
       -2.01891800e+02,  2.61862014e+02, -4.12881360e+02,  4.60696446e+01,
        1.47564153e+02,  1.87982774e+02,  3.51342702e+02, -1.60296480e+01,
       -3.42592517e+02,  2.61462426e+02, -4.99299398e+02,  1.93718698e+02,
       -1.28595796e+02,  1.04998549e+03, -3.05326957e+02,  1.10447716e+02,
       -7.07374036e+01, -

In [492]:
test_cfm_r = np.matmul(x_test, closed_form_sol(x_test, test_price))
test_cfm_r

array([ 1.58323446e+02, -2.71676554e+02, -2.83676554e+02, -1.16765540e+01,
       -2.96676554e+02,  2.13234460e+01, -2.66765540e+01, -2.16676554e+02,
       -1.67655400e+00, -2.61376554e+02, -8.16765540e+01, -6.66765540e+01,
        3.58323446e+02, -3.47655400e+00, -1.12676554e+02, -3.76868554e+02,
       -2.16765540e+01, -6.66765540e+01, -2.58676554e+02, -2.20176554e+02,
        2.98323446e+02, -2.54676554e+02, -1.66676554e+02, -1.92176554e+02,
       -9.52265540e+01, -2.26676554e+02, -2.70366554e+02, -2.26676554e+02,
       -2.96676554e+02,  1.43844600e+00,  3.58323446e+02,  1.90832345e+03,
        1.23234460e+01, -2.91776554e+02,  2.26323446e+02,  4.38323446e+02,
       -1.46676554e+02, -1.26676554e+02, -1.61676554e+02,  2.43234460e+01,
       -3.23176554e+02,  8.32344600e+00, -2.96455540e+01, -3.44176554e+02,
       -7.16765540e+01,  4.83234460e+01, -7.66765540e+01, -4.26765540e+01,
       -2.66765540e+01,  1.74823446e+02, -1.31676554e+02,  3.08323446e+02,
        5.08323446e+02,  

In [493]:
# compute r^2 score with training 
cfm_train_r_sqr_score = met.r2_score(y_true=train_price, y_pred=train_cfm_r)
# compute mean squared error
cfm_train_mse = met.mean_squared_error(train_price, train_cfm_r)

# print score and mse
print("r^2 score: " + f'{cfm_train_r_sqr_score}')
print("MSE: " + f'{cfm_train_mse}')

r^2 score: -1.6257181526734983
MSE: 302317.7672230435


In [494]:
# compute r^2 score with training 
cfm_test_r_sqr_score = met.r2_score(y_true=test_price, y_pred=test_cfm_r)
# compute mean squared error
cfm_test_mse = met.mean_squared_error(test_price, test_cfm_r)

# print score and mse
print("r^2 score: " + f'{cfm_test_r_sqr_score}')
print("MSE: " + f'{cfm_test_mse}')

r^2 score: -0.7598444173045644
MSE: 293413.48915331496


The r^2 score of the closed form solution predictor of weights on the training data is a relatively large negative value. This indicates that the predictor is very poor at predicting new test points for the desired Y response, because a negative r^2 score indicates that the predictor's predictions of all the points collectively mostly fails to fit the structure of the data. The MSE is also high, indicating very high variance of the predicted points and the actual Y values. The r^2 score on the test data is also negative and the MSE is also very high. This means that the weights produced by the closed form solution are not a good predictor of those weights.

## Problem 4

In [495]:
def poly_reg(x_feat, y_feat, p):
    '''
    Trains polynomial regression on a given training model and test point
    
    train_df (pd.DataFrame): training data DataFrame
    theta_list (list): list of predicted theta of features
    feat (list): feature class from training data
    y_feat (str): feature class to predict
    p (int): degree of polynomial model
    
    Return:
    list: list of predicted values from test point on training data
    '''
    # range of given p
    p_range = range(1, p)
    # initialize dataframe for results
    poly_x = pd.DataFrame()
    
    # range for given p to loop
    for i in range(1, p):
        # raises every x by this p
        x_raised = [ x ** i for x in x_feat ]
        # appends x_raised to dataframe
        poly_x[f'{i}'] = x_raised
        
    result = np.matmul(poly_x, closed_form_sol(poly_x, y_feat)).to_numpy()
    
    return result

In [590]:
train_poly_sqrft_df = pd.DataFrame(columns=['r^2', 'mse'])
train_poly_sqrft_df

Empty DataFrame
Columns: [r^2, mse]
Index: []

In [591]:
p_r = [2, 3, 4]

for p in p_r:
    
    train_poly = poly_reg(house_training['sqft_living'], train_price, p)
    poly_train_r_sqr = met.r2_score(y_true=train_price, y_pred=train_poly)
    # compute mean squared error
    poly_train_mse = met.mean_squared_error(train_price, train_poly)
    
    poly_row = [poly_train_r_sqr, poly_train_mse]
    
    train_poly_sqrft_df.loc[len(train_poly_sqrft_df)] = poly_row
    
train_poly_sqrft_df

<ipython-input-488-e445577b7d35>:23: FutureWarning: Calling a ufunc on non-aligned DataFrames (or DataFrame/Series combination). Currently, the indices are ignored and the result takes the index/columns of the first DataFrame. In the future , the DataFrames/Series will be aligned before applying the ufunc.
Convert one of the arguments to a NumPy array (eg 'ufunc(df1, np.asarray(df2)') to keep the current behaviour, or align manually (eg 'df1, df2 = df1.align(df2)') before passing to the ufunc to obtain the future behaviour and silence this warning.
  cfm1 = np.linalg.pinv(np.matmul(df_, matrix_t))
<ipython-input-488-e445577b7d35>:23: FutureWarning: Calling a ufunc on non-aligned DataFrames (or DataFrame/Series combination). Currently, the indices are ignored and the result takes the index/columns of the first DataFrame. In the future , the DataFrames/Series will be aligned before applying the ufunc.
Convert one of the arguments to a NumPy array (eg 'ufunc(df1, np.asarray(df2)') to keep

r^2           mse
0  4.952643e-01  5.811384e+04
1  5.053712e-01  5.695017e+04
2 -5.826129e+06  6.708041e+11

In [592]:
test_poly_sqrft_df = pd.DataFrame(columns=['r^2', 'mse'])
test_poly_sqrft_df

Empty DataFrame
Columns: [r^2, mse]
Index: []

In [593]:
for p in p_r:
    
    test_poly = poly_reg(house_testing['sqft_living'], test_price, p)
    poly_test_r_sqr = met.r2_score(y_true=test_price, y_pred=test_poly)
    # compute mean squared error
    poly_test_mse = met.mean_squared_error(test_price, test_poly)
    
    poly_row = [poly_test_r_sqr, poly_test_mse]
    test_poly_sqrft_df.loc[len(test_poly_sqrft_df)] = poly_row
    
test_poly_sqrft_df

<ipython-input-488-e445577b7d35>:23: FutureWarning: Calling a ufunc on non-aligned DataFrames (or DataFrame/Series combination). Currently, the indices are ignored and the result takes the index/columns of the first DataFrame. In the future , the DataFrames/Series will be aligned before applying the ufunc.
Convert one of the arguments to a NumPy array (eg 'ufunc(df1, np.asarray(df2)') to keep the current behaviour, or align manually (eg 'df1, df2 = df1.align(df2)') before passing to the ufunc to obtain the future behaviour and silence this warning.
  cfm1 = np.linalg.pinv(np.matmul(df_, matrix_t))
<ipython-input-488-e445577b7d35>:23: FutureWarning: Calling a ufunc on non-aligned DataFrames (or DataFrame/Series combination). Currently, the indices are ignored and the result takes the index/columns of the first DataFrame. In the future , the DataFrames/Series will be aligned before applying the ufunc.
Convert one of the arguments to a NumPy array (eg 'ufunc(df1, np.asarray(df2)') to keep

r^2           mse
0  4.658799e-01  8.905222e+04
1  5.172832e-01  8.048191e+04
2 -1.988401e+08  3.315200e+13

As the degree of the polynomial increases for both the training and testing data, the r^2 score also decreases, to the point that it is negative. This indicates that my polynomial regression model does not adhere to the principles of polynomial regression, which should in theory increase the r^2 score because it should be better at reducing the bias in the data. The MSE is decreases as the polynomial degree increases. This is in line with the idea that polynomial regression should reduce the variance and bias of the model compared to linear regression.

## Problem 5

fix iter looping add theta

In [566]:
def gradient_descent(theta, y_, x_, iters, alpha):
    '''
    Returns the adjusted theta after gradient descent is applied
    
    theta (array): matrix of theta weights derived from model
    y_ (array): 1-d array of response feature
    x_ (array): training dataset
    iters (num): number of iterations
    
    Returns:
    (array): new theta weights after GD update
    '''
    
    theta_prime = theta
    
    for ith in range(1, iters):
        
        error_term = np.subtract(np.matmul(x_, theta), y_)
        
        trans_term = np.transpose(error_term).tolist()
    
        sub_array = np.matmul(trans_term, x_)
        
        s_array = np.dot(2, sub_array)
        
        su_array = np.divide(s_array, x_.size)
    
        next_theta = np.dot(su_array, alpha)
        
        theta_prime = np.add(theta_prime, next_theta)
        
    results = np.matmul(x_, theta_prime)
        
    #return s_array
    #return theta_prime
    #return error_term
    #return trans_term
    return results, theta_prime

In [611]:
x_train

array([[ 0.        , -0.40982347, -1.44988843, ..., -0.35519332,
        -0.96563661, -0.3128578 ],
       [ 0.        , -0.40982347,  0.28318404, ..., -0.7998304 ,
        -0.44332966, -0.23355563],
       [ 0.        , -1.58410276, -1.44988843, ..., -0.18307573,
         1.09374507, -0.21669046],
       ...,
       [ 0.        , -1.58410276, -1.44988843, ..., -0.86437449,
        -1.02532883, -0.4185143 ],
       [ 0.        , -1.58410276, -1.44988843, ...,  1.07911986,
        -0.80148299, -0.40352304],
       [ 0.        , -0.40982347, -0.06343045, ..., -0.46993837,
         0.39236146, -0.15736334]])

In [610]:
len(gradient_descent(train_cfm, train_price, x_train, 10, 0.01)[1])

18

In [609]:
len(gradient_descent(train_cfm, train_price, x_train, 10, 0.01)[0])

1000

In [558]:
train_gd_df = pd.DataFrame(columns=['gd', 'r^2', 'MSE'])
train_gd_df

Empty DataFrame
Columns: [gd, r^2, MSE]
Index: []

In [568]:
alpha_list = [0.001, 0.01, 0.1]

iter_list = [10, 50, 100]

for it in iter_list:
    
    for al in alpha_list:
        
        train_gd, thetas = gradient_descent(train_cfm, train_price, x_train, it, al)
        
        r2 = met.r2_score(y_true=train_price, y_pred=train_gd)
        
        # compute mean squared error
        mse_ = met.mean_squared_error(train_price, train_gd)
        
        gd_row = ['iters=' + f'{it}' + ' a=' + f'{al}', r2, mse_]
        
        train_gd_df.loc[len(train_gd_df)] = gd_row
        
train_gd_df

gd       r^2            MSE
9    iters=100 a=0.1 -1.625718  302317.767223
1   iters=10 a=0.001 -1.625718  302317.767223
2    iters=10 a=0.01 -1.625718  302317.767223
3     iters=10 a=0.1 -1.625718  302317.767223
4   iters=50 a=0.001 -1.625718  302317.767223
5    iters=50 a=0.01 -1.625718  302317.767223
6     iters=50 a=0.1 -1.625718  302317.767223
7  iters=100 a=0.001 -1.625718  302317.767223
8   iters=100 a=0.01 -1.625718  302317.767223

In [569]:
test_gd_df = pd.DataFrame(columns=['gd', 'r^2', 'MSE'])
test_gd_df

Empty DataFrame
Columns: [gd, r^2, MSE]
Index: []

In [570]:
alpha_list = [0.001, 0.01, 0.1]

iter_list = [10, 50, 100]

for it in iter_list:
    
    for al in alpha_list:
        
        test_gd, thetas = gradient_descent(test_cfm, test_price, x_test, it, al)
        
        r2 = met.r2_score(y_true=test_price, y_pred=test_gd)
        
        # compute mean squared error
        mse_ = met.mean_squared_error(test_price, test_gd)
        
        gd_row = ['iters=' + f'{it}' + ' a=' + f'{al}', r2, mse_]
        
        test_gd_df.loc[len(test_gd_df)] = gd_row
        
test_gd_df

gd       r^2            MSE
0   iters=10 a=0.001 -0.759844  293413.489153
1    iters=10 a=0.01 -0.759844  293413.489153
2     iters=10 a=0.1 -0.759844  293413.489153
3   iters=50 a=0.001 -0.759844  293413.489153
4    iters=50 a=0.01 -0.759844  293413.489153
5     iters=50 a=0.1 -0.759844  293413.489153
6  iters=100 a=0.001 -0.759844  293413.489153
7   iters=100 a=0.01 -0.759844  293413.489153
8    iters=100 a=0.1 -0.759844  293413.489153

My gradient descent model seems to produce the same, if not similar values for r^2 score and MSE for both the training and testing datasets regardless of how the alpha and iterations are varied. This may make sense, as my closed-form solution may be such as good predictor that gradient descent does not improve the quality of the model much, or that the changes are too small for it to be observed using these numbers. Gradient descent should increase the r^2 and decrease the MSE, because it optimizes the weights for each feature by gradually adjusting the weights based on the formula. 

## Problem 6

Set the gradient equal to 0 to solve

In [665]:
x_train

array([[ 0.        , -0.40982347, -1.44988843, ..., -0.35519332,
        -0.96563661, -0.3128578 ],
       [ 0.        , -0.40982347,  0.28318404, ..., -0.7998304 ,
        -0.44332966, -0.23355563],
       [ 0.        , -1.58410276, -1.44988843, ..., -0.18307573,
         1.09374507, -0.21669046],
       ...,
       [ 0.        , -1.58410276, -1.44988843, ..., -0.86437449,
        -1.02532883, -0.4185143 ],
       [ 0.        , -1.58410276, -1.44988843, ...,  1.07911986,
        -0.80148299, -0.40352304],
       [ 0.        , -0.40982347, -0.06343045, ..., -0.46993837,
         0.39236146, -0.15736334]])

In [677]:
def append_ridge(theta, y_, x_, iters, alpha, lam):
    '''
    Appends ridge regression term to given theta list 
    
    theta (list): 1-d array of theta values
    y_ (list): 1-d array of y to fit gradient descent
    x_ (array): 2-d array of training data for gradient descent
    iters (int): number of iterations of gradient descent
    alpha (float): alpha factor
    lam (float): lambda control factor
    
    Return:
    (array): updated training dataset with adjusted weights
    
    '''
    # uses gradient descent to produce array of updated dataset and thetas
    new_vals, adj_thetas = gradient_descent(theta, y_, x_, iters, alpha)
    
    # ridge value
    ridge_ = alpha * lam
    # multiplies ridge by each theta
    ridge = np.dot(adj_thetas, ridge_)
    
    ridge_appd = np.add(adj_thetas, ridge)
    
    x_scaled = np.reshape(x_, (-1, len(x_[0])))
    
    rdg_result = np.matmul(x_scaled, ridge_appd)
    '''
    result_scaled = np.reshape(rdg_result, (-1, 18))
    '''
    return rdg_result
    #return result_scaled
    #return x_scaled

In [678]:
append_ridge(train_cfm, train_price, x_train, 100, 0.001, 4)

array([-2.50004305e+02,  2.09538465e+02, -9.31286797e+01, -9.45919623e+01,
       -7.02179655e+01,  8.63817652e+02, -2.81662820e+02, -3.06946875e+02,
       -1.56718849e+02, -2.32210528e+02,  2.49931128e+02, -8.89185495e+01,
        4.30808343e+01, -1.84896841e+02,  5.65060272e+01,  4.05094243e+02,
       -5.89159421e+01,  1.33101091e+01, -2.61908795e+02, -3.58203280e+02,
       -4.15772774e+01,  6.18322369e+02, -1.47124989e+02, -4.04999463e+02,
       -3.47206168e+01, -2.94195985e+02,  2.20272857e+02,  1.10898086e+02,
       -5.70177190e+01,  1.19804871e+02,  1.06590982e+01, -1.85701311e+02,
        9.18027923e+01,  2.25443694e+01, -1.74486889e+02,  2.77356726e+02,
       -2.02699368e+02,  2.62909462e+02, -4.14532886e+02,  4.62539232e+01,
        1.48154410e+02,  1.88734705e+02,  3.52748073e+02, -1.60937666e+01,
       -3.43962887e+02,  2.62508276e+02, -5.01296596e+02,  1.94493573e+02,
       -1.29110179e+02,  1.05418543e+03, -3.06548265e+02,  1.10889506e+02,
       -7.10203532e+01, -

In [573]:
rand_x_list = np.random.normal(-2, 2, 1000)
rand_e_list = np.random.normal(0, 1, 1000)

In [634]:
norm_col_p6 = [1] * len(rand_x_list)

In [636]:
prob6_df = pd.DataFrame()

In [637]:
prob6_df['x'] = rand_x_list

In [638]:
prob6_df.insert(loc=0, column='norm', value=norm_col)

In [648]:
prob6_df

norm         x
0       1 -7.273312
1       1 -7.650463
2       1 -1.478170
3       1 -2.206831
4       1 -2.291508
..    ...       ...
995     1 -5.161616
996     1 -0.880706
997     1 -3.662544
998     1 -2.053737
999     1 -2.677474

[1000 rows x 2 columns]

In [656]:
prob6_array = prob6_df.to_numpy()
prob6_array

array([[ 1.        , -7.27331213],
       [ 1.        , -7.65046275],
       [ 1.        , -1.47816956],
       ...,
       [ 1.        , -3.66254443],
       [ 1.        , -2.05373686],
       [ 1.        , -2.67747406]])

In [574]:
y_list = []

for x_, e_ in zip(rand_x_list, rand_e_list):
    
    y_ = 1 + 2 * x_ + e_
    y_list.append(y_)
    
y_list

[-12.86365708640485,
 -14.440902690676605,
 -2.269989112511947,
 -3.325456732650936,
 -5.269266788449755,
 2.4342352066796495,
 -10.035911451447092,
 -5.440582909961237,
 -5.354335122816812,
 -7.232705008675326,
 -9.645665677191865,
 1.6425142274720335,
 1.9670168111071253,
 -8.666967923321517,
 3.2668272437883146,
 -2.48465664366042,
 -3.6537689845215753,
 8.126051500375308,
 -5.963699616309118,
 -3.7196570861256286,
 -0.9406455933222351,
 -2.665282092908721,
 -1.649043461534267,
 -6.898723555732141,
 -9.688878884052409,
 -3.006447232470402,
 0.574618071234084,
 2.7693527508465827,
 -7.6640351006567355,
 -10.93313404446484,
 -5.9353635640630635,
 3.4587836901638602,
 1.0130268443623678,
 -0.21743379682945274,
 -10.09242340459523,
 3.5807530022283363,
 -7.348436106914018,
 -1.9455545757453523,
 -5.349617183798018,
 -7.446193537519775,
 -7.984369048688716,
 1.3674840956594427,
 -7.581212940718351,
 -1.6543115684728076,
 -3.447318508257754,
 3.9075643265667446,
 0.08057735517887854,
 -5.

In [654]:
# remove mean and fit to variance
train_scaler1 = StandardScaler()
# fit training data to scaler
#train_scaler1.fit(rand_x_list)
# perform standardization by scaling
x_train_rdg = pd.DataFrame()

x_train_rdg['x_'] = rand_x_list
x_train_rdg['y_'] = y_list
# training set response y
y_train = y_list

In [625]:
len(y_train)

1000

In [626]:
len(rand_x_list)

1000

In [686]:
# test closed_form_sol on training set and training price
rdg_cfm = closed_form_sol(x_train_rdg, y_train)
rdg_cfm = rdg_cfm.to_numpy()
rdg_cfm

<ipython-input-488-e445577b7d35>:23: FutureWarning: Calling a ufunc on non-aligned DataFrames (or DataFrame/Series combination). Currently, the indices are ignored and the result takes the index/columns of the first DataFrame. In the future , the DataFrames/Series will be aligned before applying the ufunc.
Convert one of the arguments to a NumPy array (eg 'ufunc(df1, np.asarray(df2)') to keep the current behaviour, or align manually (eg 'df1, df2 = df1.align(df2)') before passing to the ufunc to obtain the future behaviour and silence this warning.
  cfm1 = np.linalg.pinv(np.matmul(df_, matrix_t))


array([-3.09136780e-04,  9.99206239e-01])

In [628]:
# initializes Linear Regression
reg = LinearRegression()

# fits the training
reg1 = reg.fit(x_train_rdg, y_train)

#for column in house_train_n:
pred_test = reg1.predict(x_train_rdg)
#y_pred_list.append(pred_y)
    
#y_pred_list
print(reg.coef_)

[1.10572273e-15 1.00000000e+00]


In [629]:
# compute r^2 score with training 
train_r_sqr_score = met.r2_score(y_true=y_train, y_pred=pred_test)
# compute mean squared error
train_mse = met.mean_squared_error(y_train, pred_test)

# print score and mse
print("r^2 score: " + f'{test_r_sqr_score}')
print("MSE: " + f'{test_mse}')

r^2 score: 1.0
MSE: 8.691778000985667e-26


In [685]:
rdg_cfm

array([-3.09136780e-04,  9.99206239e-01])

In [683]:
y_train

[-12.86365708640485,
 -14.440902690676605,
 -2.269989112511947,
 -3.325456732650936,
 -5.269266788449755,
 2.4342352066796495,
 -10.035911451447092,
 -5.440582909961237,
 -5.354335122816812,
 -7.232705008675326,
 -9.645665677191865,
 1.6425142274720335,
 1.9670168111071253,
 -8.666967923321517,
 3.2668272437883146,
 -2.48465664366042,
 -3.6537689845215753,
 8.126051500375308,
 -5.963699616309118,
 -3.7196570861256286,
 -0.9406455933222351,
 -2.665282092908721,
 -1.649043461534267,
 -6.898723555732141,
 -9.688878884052409,
 -3.006447232470402,
 0.574618071234084,
 2.7693527508465827,
 -7.6640351006567355,
 -10.93313404446484,
 -5.9353635640630635,
 3.4587836901638602,
 1.0130268443623678,
 -0.21743379682945274,
 -10.09242340459523,
 3.5807530022283363,
 -7.348436106914018,
 -1.9455545757453523,
 -5.349617183798018,
 -7.446193537519775,
 -7.984369048688716,
 1.3674840956594427,
 -7.581212940718351,
 -1.6543115684728076,
 -3.447318508257754,
 3.9075643265667446,
 0.08057735517887854,
 -5.

In [690]:
rdg_reg = append_ridge(rdg_cfm, y_train, prob6_array, 100, 0.001, 1)
np.reshape(rdg_reg, (-1, 2)

500

In [680]:
lambda_list = [1, 10, 100, 1000, 10000]

rdg_df = pd.DataFrame(columns=['x_', 'y_'])

rdg_list = []

for lam in lambda_list:
    rdg_reg = append_ridge(rdg_cfm, y_train, prob6_array, 100, 0.001, lam)
    #rdg_df.loc[len(rdg_df)] = gd_reg_thetas
    rdg_list.append(rdg_reg)

In [681]:
rdg_list

[array([-2.68046740e+00, -2.82475555e+00, -4.63394602e-01, -7.42161742e-01,
        -7.74556917e-01,  2.06340553e-01, -1.81883889e+00, -9.04035749e-01,
        -9.30591642e-01, -1.53306928e+00, -1.91356466e+00,  2.46599689e-01,
         2.82913456e-01, -1.64739334e+00,  3.13180666e-01, -7.31447528e-01,
        -8.94008054e-01,  1.24279598e+00, -1.30273234e+00, -7.55474966e-01,
        -2.95983559e-01, -9.32036916e-01, -7.28269943e-01, -1.31401436e+00,
        -2.21644472e+00, -4.47010489e-01, -1.47182699e-02,  4.87006734e-01,
        -1.45661192e+00, -2.14362562e+00, -1.14609744e+00,  4.35050247e-01,
        -4.18960199e-02,  1.61717054e-02, -1.90910506e+00,  7.76575404e-01,
        -1.32162796e+00, -5.37126622e-01, -1.30382825e+00, -1.08332098e+00,
        -1.20082288e+00, -2.26306057e-01, -1.52543847e+00, -4.73498698e-01,
        -7.92353535e-01,  6.44241824e-01,  9.97423867e-02, -8.51000843e-01,
        -1.06158416e+00, -6.06562915e-01, -7.64310655e-01, -1.84085568e+00,
        -5.3

In [443]:
linregress(rdg_df['x_'], rdg_df['y_'])

LinregressResult(slope=1472.058340643426, intercept=-3.0803910659023037, rvalue=1.0, pvalue=1.2004217548761408e-30, stderr=0.0, intercept_stderr=0.0)

In [445]:
# compute r^2 score with training 
rdg_r_sqr_score = met.r2_score(y_true=y_train, y_pred=rdg_df['y_'])
# compute mean squared error
rdg_mse = met.mean_squared_error(y_train, rdg_df['y_'])

# print score and mse
print("r^2 score: " + f'{rdg_r_sqr_score}')
print("MSE: " + f'{rdg_mse}')

ValueError: Found input variables with inconsistent numbers of samples: [1000, 5]